contracts/src/TeleporterTokenDestination.sol

```diff
@@ -0,0 +1,85 @@
+import {IERC20} from "@openzeppelin/contracts@4.8.1/token/ERC20/IERC20.sol";
+
+abstract contract TeleporterTokenDestination is TeleporterOwnerUpgradeable {
+    address public constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+    bytes32 public immutable blockchainID;
+
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable teleporterTokenSourceAddress;
+    IERC20 public immutable token;
+
+    event SendTokens(bytes32 indexed teleporterMessageID, address indexed sender, uint256 amount);
+
+    constructor(
+        address teleporterRegistryAddress,
+        address teleporterManager,
+        bytes32 sourceBlockchainID_,
+        address teleporterTokenSourceAddress_,
+        address token_
+    ) TeleporterOwnerUpgradeable(teleporterRegistryAddress, teleporterManager) {
+        sourceBlockchainID = sourceBlockchainID_;
+        teleporterTokenSourceAddress = teleporterTokenSourceAddress_;
+        token = IERC20(token_);
+        blockchainID = IWarpMessenger(WARP_PRECOMPILE_ADDRESS).getBlockchainID();
+    }
+
+    function _send(
+        bytes32 destinationBlockchainID,
+        address destinationBridgeAddress,
+        address recipient,
+        uint256 amount,
+        uint256 primaryFee,
+        uint256 secondaryFee
+    ) internal virtual {
+        // increase bridge balance by amount
+        // send message to destinationBridgeAddress
+        bytes32 messageID = _sendTeleporterMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: destinationBlockchainID,
+                destinationAddress: destinationBridgeAddress,
+                feeInfo: TeleporterFeeInfo({feeTokenAddress: address(token), amount: primaryFee}),
+                requiredGasLimit: 0,
+                allowedRelayerAddresses: new address[](0),
+                message: abi.encode(
+                    destinationBlockchainID, destinationBridgeAddress, recipient, amount, secondaryFee
+                    )
+            })
+        );
+
+        emit SendTokens(messageID, msg.sender, amount);
+    }
+
+    function _receiveTeleporterMessage(
+        bytes32 sourceBlockchainID_,
+        address originSenderAddress,
+        bytes memory message
+    ) internal virtual override {
+        // Only allow messages from the source chain.
+        require(
+            sourceBlockchainID_ == sourceBlockchainID,
+            "TeleporterTokenDestination: invalid source chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(
+            originSenderAddress == teleporterTokenSourceAddress,
+            "TeleporterTokenDestination: invalid Teleport token source address"
+        );
+        (address recipient, uint256 amount) = abi.decode(message, (address, uint256));
+
+        _withdraw(recipient, amount);
+    }
+
+    function _withdraw(address recipient, uint256 amount) internal virtual {}
```

Comment:
Can we name this function `mint` on the destination and `unlock` on the source? I think that would be a lot more intuitive as to what we're doing.

Reply:
I'd prefer to revisit this after the native and erc20 token bridge contracts, want to see how applicable deposit/withdraw functions are to each of them, and if there are any possible abstractions between their implementations